keps/sig-auth/20190916-noderestriction-pods.md

```diff
@@ -0,0 +1,476 @@
+
+### Background
+
+The Kubelet has two mechanisms that can be used to manipulate pods. The first is through the
+pod/status subresource. Despite the name, an update to pod/status can also update (some) of the
+pod's metadata. Of particular interest are updates to labels and annotations. Note that pod/status
+updates _cannot_ modify the OwnerReferences [[1][]].
+
+The second mechanism is through the creation of "mirror pods". The Kubelet can run pods from other
+sources than the API server, such as a static manifest directory or pulled from an HTTP
+server. These pods are referred to as "static pods". Since static pods don't come from the API
+server, components that read pods from the API wouldn't know about them. To compensate, the Kubelet
+creates a "mirror pod", which reflects the state of the static pod it's running.
+
+Mirror pods have some special properties. They are identified with a special annotation,
+`kubernetes.io/config.mirror`, and the Kubelet is only authorized to create mirror pods, and only on
+the same node (itself). The Kubelet won't run a mirror pod (since it's actually running a static
+pod). Mirror pods are also restricted from using a service account, or secrets, configmaps,
+persistent volumes, and other resources restricted by the node authorizer, in order to prevent an
+attacker from bypassing the authorization by creating mirror pods.
+
+[1]: https://github.com/kubernetes/kubernetes/blob/ab73a018de51bddf9d03d6fed6e867b60196c796/pkg/registry/core/pod/strategy.go#L162-L171
+
+[Node Authorizer]: https://kubernetes.io/docs/reference/access-authn-authz/node/
+[NodeRestriction]: https://kubernetes.io/docs/reference/access-authn-authz/node/
+[man-in-the-middle]: https://en.wikipedia.org/wiki/Man-in-the-middle_attack
+
+### Threat Model
+
+At a high level, this proposal targets clusters making effective use of node isolation to separate
+sensitive workloads or limit the blast radius of a successful node compromise. More specifically, it
+makes the following assumptions:
+
+- The cluster uses scheduling constraints such as node selectors & taints / tolerations to separate
+  workloads of different trust or privilege levels.
+- An attacker has compromised a low-privilege node meaning they have code execution as root in the
+  host namespaces (i.e. a container escape). Low-privilege in this case means the node is not
+  hosting any privileged workloads with permissions to trivially take over the cluster
+  (e.g. unrestricted pod creation, read secrets, etc).
+
+### Goals
+
+- Prevent a compromised node from manipulating controllers to execute confused deputy attacks.
+
+### Non-Goals
+
+- Solving all node isolation issues.
+- Solving all possible man-in-the-middle attacks.
+
+## Proposal
+
+### Node Restrictions
+
+All restrictions will be enforced through the [NodeRestriction][] admission controller. These
+extensions will be guarded by the `NodeRestrictionPods` feature gate.
```

Comment:
some restrictions for things the kubelet does not do today can be added immediately, right?

suggest gating the restrictions specific to mirror pod labels/ownerReferences and naming it something specific like `MirrorPodMetadataNodeRestriction`

Reply:
Yeah, makes sense. So to clarify, you're proposing making the `pod/status` update label restrictions un-gated?